### PR TITLE
Add coverage testing on Jenkins

### PR DESCRIPTION
**Pull request by Pansysk75**
Runs coverage tests on Jenkins, and uploads results to Codacy.
Relies on fetching `grcov` utility.

### CODE REVIEW

**Comment by hkaiser** [comment]
Please ignore the cmake formatting issue, that one has trickled through from master.

**Comment by Pansysk75** [comment]
retest lsu

**Comment by Pansysk75** [comment]
> ## Coverage summary from Codacy
> ...

Finally this worked :) I will disable this guy, we don't need a comment.

I'll add the ReadMe badge as well, and check whether I forgot to include some part of HPX in the build.

Currently, it runs for every PR, which takes 2 to 3 hours on Rostam (because it builds and runs all tests). I think that's fine for now, but we could get away only running it on master, to save resources.

Also, currently, it reports as failed if some build/run failed, but it will intentionally still attempt to get the coverage results.

File: README.rst [hunk 1]
```diff
@@ -151,13 +151,13 @@ Past and current funding and support for HPX is listed `here
...
-.. |codacy| image:: https://app.codacy.com/project/badge/Grade/0b8cd5a874914edaba67ce3bb711e688
+.. |codacy_quality| image:: https://app.codacy.com/project/badge/Grade/0b8cd5a874914edaba67ce3bb711e688
      :target: https://app.codacy.com/gh/STEllAR-GROUP/hpx/dashboard
      :alt: HPX Code Quality Assessment
 
-.. |coveralls| image:: https://coveralls.io/repos/github/STEllAR-GROUP/hpx/badge.svg
-     :target: https://coveralls.io/github/STEllAR-GROUP/hpx
-     :alt: HPX coverage report
+.. |codacy_coverage| image:: https://app.codacy.com/project/badge/Coverage/0b8cd5a874914edaba67ce3bb711e688    
+  :target: https://app.codacy.com/gh/STEllAR-GROUP/hpx/dashboard?utm_source=gh&utm_medium=referral&utm_content=&utm_campaign=Badge_coverage
```
Comment (hkaiser) [inline README.rst:159]: I think you can remove all the `?key=value&...` parameters on this url.

Comment (Pansysk75) [reply README.rst:159]: Codacy gave this to me as-is, but you are completely right in that it doesn't make a single difference.

**Comment by Pansysk75** [comment]
@hkaiser This is ready to merge, the only caveat being that things like APEX and CUDA (and probably others) aren't included in the tests that are currently being ran, so those parts of the codebase might not show up as "covered" by tests. Same goes for compile time tests.

I suggest merging this first so that we can a few runs and thus get a better image in the dashboard. We can tweak it later if needed.

**Comment by hkaiser** [comment]
> @hkaiser This is ready to merge, the only caveat being that things like APEX and CUDA (and probably others) aren't included in the tests that are currently being ran, so those parts of the codebase might not show up as "covered" by tests. Same goes for compile time tests.
> 
> I suggest merging this first so that we can a few runs and thus get a better image in the dashboard. We can tweak it later if needed.

Thanks @Pansysk75! This is great. 

> Oh, also @hkaiser you'll need to disable the commenting because you are the admin in Codacy. It's in the settings tab, under "Integrations". Thanks!

Should we really disable the comments?

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
bors merge

**Comment by Pansysk75** [comment]
> Thanks @Pansysk75! This is great.

My pleasure :))
 
> Should we really disable the comments?

We can give them a chance, I don't mind at all. You can always toggle them directly from the dashboard.

**Pull request merged**